### PR TITLE
common,runtime, babe, core: Runtime Call Safety

### LINKED ISSUE #359: Runtime & Concurrency
I think it's inevitable that the runtime will be accessed by concurrent processes. We should enforce a mutex within the available calls, including memory accesses. There is no evidence that wasmer is thread safe.

On the converse, it seems unnecessary for the allocator to require a lock as Wasm is single threaded. We can probably remove this from the `ext_` functions without any risk.

### PR DESCRIPTION
**Pull request by dziabko**
## Changes
- Made PriorityQueue calls thread-safe
- Added lock to runtime Exec calls
- Made runtime Exec call accept arbitrary data, parses it & calls runtime

## Tests:
```
 go test -race ./runtime
```

### Issues:
closes #359

### CODE REVIEW

File: common/transaction/priority_queue.go [hunk 1]
```diff
@@ -27,6 +30,16 @@ type node struct {
...
+		head:  nil,
+		mutex: sync.Mutex{},
+	}
+
+	return &pq
+}
+
 // Pop removes the head of the queue and returns it
 func (q *PriorityQueue) Pop() *ValidTransaction {
 	head := q.head
```
Comment (ansermino) [inline common/transaction/priority_queue.go]: Need lock here

File: common/transaction/priority_queue.go [hunk 2]
```diff
@@ -47,7 +62,9 @@ func (q *PriorityQueue) Peek() *ValidTransaction {
 func (q *PriorityQueue) Insert(vt *ValidTransaction) {
 	curr := q.head
 	if curr == nil {
+		q.mutex.Lock()
 		q.head = &node{data: vt}
+		q.mutex.Unlock()
 		return
 	}
```
Comment (ansermino) [inline common/transaction/priority_queue.go]: Need to protect the code below

File: runtime/wasmer_test.go [hunk 3]
```diff
@@ -892,3 +893,22 @@ func TestExt_secp256k1_ecdsa_recover(t *testing.T) {
...
+	// Execute 2 concurrent calls to the runtime
+	go func() {
+		_, _ = r.Exec("Core_version", 1, 1)
+	}()
+	go func() {
+		_, _ = r.Exec("Core_version", 1, 1)
+	}()
+
+	// Wait for routines to return runtime calls
+	time.Sleep(2 * time.Second)
```
Comment (ansermino) [inline runtime/wasmer_test.go]: Please replace this with a `WaitGroup`

File: runtime/wasmer.go [hunk 4]
```diff
@@ -115,12 +119,27 @@ func (r *Runtime) Load(location, length int32) []byte {
 	return mem[location : location+length]
 }
 
-func (r *Runtime) Exec(function string, data, len int32) ([]byte, error) {
+func (r *Runtime) Exec(function string, loc int32, data interface{}) ([]byte, error) {
+	r.mutex.Lock()
+
+	// Find out the type of data & store it correctly into memory
+	var leng int32 = 0
+	switch data.(type) {
```
Comment (noot) [inline runtime/wasmer.go]: not sure if this is needed, just passing in the length might be simpler than having to keep track of all the different possible types here

**Pull request merged**